### ISSUE TITLE
Tracking Issue for #82703 #![feature(nonzero_ops)]

### ISSUE THREAD

**Issue by iago-lito** [rust-lang/rust#84186]
Feature gate: `#![feature(nonzero_ops)]`

This is the tracking issue for #82703. The common arithmetic operators that keep the invariants of `NonZero*` types enforced are implemented for these types.

### Public API

Example with `U8`/`I8`, but same for `U16`/`I16` *etc.*
```rust
use std::num::{NonZeroU8, NonZeroI8};

impl NonZeroU8 {
    pub fn checked_add(self, other: U8) -> Option<Self>;
    pub fn saturating_add(self, other: U8) -> Self;
    pub unsafe fn unchecked_add(self, other: U8) -> Self;
    pub fn checked_next_power_of_two(self) -> Option<Self>;
}

impl NonZeroI8 {
    pub fn abs(self) -> Self;
    pub fn checked_abs(self) -> Option<Self>;
    pub fn overflowing_abs(self) -> (Self, bool);
    pub fn saturating_abs(self) -> Self;
    pub fn wrapping_abs(self) -> Self;
    pub fn unsigned_abs(self) -> U8;
}

impl NonZeroU8 /* and NonZeroI8 */ {
    pub fn checked_mul(self, other: Self) -> Option<Self>;
    pub fn saturating_mul(self, other: Self) -> Self;
    pub unsafe fn unchecked_mul(self, other: Self) -> Self;
    pub fn checked_pow(self, other: u32) -> Option<Self>;
    pub fn saturating_pow(self, other: u32) -> Self;
}
```

### Steps / History

- [x] Implementation: #82703
- [ ] Final commenting period (FCP)
- [ ] Stabilization PR
- [ ] Add `impl Neg for NonZeroI*`, which could not be feature-gated.

### Unresolved Questions

- None yet.

**Comment by iago-lito** [comment]
> -> Leave as-is (non-const) until rust-lang/rfcs#3016 lands. (#82703 (comment))

How to not forget re-checking the constness of unchecked methods when this happens? Is there a way to leave some kind of mark in the repo that'll automatically trigger this discussion again when it's time?

**Comment by Nilstrieb** [comment]
I just had a use for this, what's the state of this feature?

**Comment by iago-lito** [comment]
Hi @Nilstrieb @jschwe. The state is that I don't quite know what's next. IIUC someone needs to write a stabilisation report for the feature to move forward, but this requires that a few use cases be collected first, right? I don't have many on my side, since I think I've only used it once then, but I'm happy to help :)

**Comment by jschwe** [comment]
@iago-lito Is `nonzero_ops` not just a "library feature"? Looking at the guide for [Stabilizing a library feature](https://rustc-dev-guide.rust-lang.org/stability.html#stabilizing-a-library-feature) a stabilisation report does not seem to be necessary.

**Comment by iago-lito** [comment]
Hi @m-ou-se, is this the right time to ask for a FCP on this feature?

**Comment by m-ou-se** [comment]
I've updated the [top comment](https://github.com/rust-lang/rust/issues/84186#issue-857800515) with an up to date summary of this feature.

@rfcbot merge

**Comment by dtolnay** [comment]
It's not clear to me from the stabilization proposal, but presumably we're not stabilizing `unchecked_add` and `unchecked_mul` as part of this? That is not stable on the primitive integers yet and has a different tracking issue: https://github.com/rust-lang/rust/issues/85122.

**Comment by iago-lito** [comment]
@dtolnay I don't remember that two methods here were not supposed to be stabilized along with the others. How come I didn't have to use the *e.g.* `unchecked_math` feature gate to implement these if *e.g.* `u8::unchecked_add` were not stabilized?

**Comment by andylizi** [comment]
> How come I didn't have to use the _e.g._ `unchecked_math` feature gate to implement these

Presumably because `unchecked_math` is also defined in `libcore`, and

https://github.com/rust-lang/rust/blob/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/compiler/rustc_middle/src/middle/stability.rs#L385-L389

**Comment by iago-lito** [comment]
Hm, I think I see. So, to answer your question @dtolnay : no, this was not at all anticipated, at least on my side. What are the options then?
- Should we wait before stabilization of #85122 before stabilizing this?
- Should we only stabilize the other methods, and then add another feature gate / tracking issue just for the two remaining `unchecked_{add/mul}` methods?
- Should we transfer the two `unchecked_{add/mul}` methods to #85122?

**Comment by jschwe** [comment]
As a user, I would prefer to at least have the checked methods available on stable, since this is currently all that is preventing me from using the types.

**Comment by Nilstrieb** [comment]
I agree, stabilizing the checked methods would be useful right now, we can always stabilize unchecked later.

**Comment by TennyZhuang** [comment]
How about splitting them to `nonzero_checked_ops` and `nonzero_unchecked_ops`?

**Comment by iago-lito** [comment]
That's a good idea @TennyZhuang :) I'm just back from vacations, I'll do this as soon as I emerge from the global catchup.

**Comment by scottmcm** [comment]
Note that splitting the feature causes some churn for any users on nightly.

You might consider a PR to stabilize the checked ops as `nonzero_checked_op` -- since there's no churn to add that to `feature`s since they're stable -- while leaving the `unchecked` ones under the previous feature name.  Then either that could be the thing that gets FCP'd, or used to answer the question about specifically what's being stabilized.

See the upcoming text in https://github.com/rust-lang/std-dev-guide/pull/32#discussion_r825018680 for more details.

**Comment by yaahc** [comment]
moving the FCP to the new partial stabilization

@rfcbot cancel

**Comment by Apersoma** [comment]
Why are there no strict variants?

**Comment by iago-lito** [comment]
What do you mean @Apersoma?

**Comment by nxsaken** [comment]
@iago-lito see this: https://github.com/rust-lang/rust/issues/118260

**Comment by iago-lito** [comment]
I see, well I guess I had no idea about these variants back in 2021 or until today :) I am not sure what the procedure is. Should we open another PR/issue to discuss possible integration within `nonzero_ops`, or a possible new `nonzero_strict_ops` instead?

This whole initiative started even before [`NonZero<T>`](https://doc.rust-lang.org/std/num/struct.NonZero.html). Maybe this all needs refresh.

**Comment by Apersoma** [comment]
@iago-lito 
Since the FCP appears to be over, I'm not sure.
